### PR TITLE
Reproducibility, linalg speedup, renaming

### PR DESCRIPTION
**Pull request by tommyod**
- Using `random_state` as input along with a `check_random_state()` function to facilitate reproducibility. This is how sklearn and other packages typically do it.
- Replaced the `len(array.shape)` checks with `array.ndim`. Hopefully a bit more readable
- Renamed a variable
- Avoid forming two diagonal matrices to pre-and-post multiply, instead just multiply each row/column. Goes from `O(m^3)` to `O(m^2)`. Probably not a big practical speedup, since this is unlikely to be a bottleneck, but might as well optimize these kinds of operations in my opinion.

This is a first step toward getting rid of `noise`, which I will do in another PR

I chose to just pass an integer `seed` instead of allowing a random state object, after reading
- https://github.com/scikit-learn/scikit-learn/issues/14042
- https://github.com/scikit-learn/scikit-learn/issues/16988
- https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.default_rng

I don't see a usecase where an integer seed is not good enough.

### CODE REVIEW

**Review by dafeda — APPROVED** [review]
Very nice 👍
